### ISSUE TITLE
Bug: useEffect doesn't work asynchronously

### ISSUE THREAD

**Issue by NiedziolkaMichal** [facebook/react#24982]
React version: 18.2.0

## Steps To Reproduce

1. Add TestComponent to your App
2. Run App with DevTools console open
3. Watch exactly when each log is shown and when App is displayed on the screen
4. Click on button
5. Watch exactly when displayed "Current State" changes and when each log is shown

Link to code example:
```
import React, {useState, useEffect, useLayoutEffect} from 'react'

export function TestComponent(props) {
    const [count, setCount] = useState(0);

    useEffect(() => {
        console.log('useEffect start, buttonState=', document.getElementById('test-btn').className);
        let waitOverEpoch = new Date().getTime() + 2000;
        while(waitOverEpoch > new Date().getTime())
            ;
        console.log('useEffect over');
    });

    useLayoutEffect(() => {
        console.log('useLayoutEffect, buttonState=', document.getElementById('test-btn').className);
    });

    function handleClick() {
        console.log('Clicked button');
        setCount(count + 1);
        console.log('Scheduling Timeout');

        setTimeout(() => {
            console.log('Timeout is done. Changing state');
            setCount(c => c + 1)
        }, 10000);
    }

    console.log('render');
    return <div>
        <p>Current State: {count}</p>
        <button id="test-btn" className={"count_" + (count)} onClick={handleClick}>Increase Count</button>
    </div>
}
```

## The current behavior
When app is being loaded for the first time, browser displays it only after "useEffect over" message is shown on the dev logs. Same thing happens after clicking on button and after timeout.

## The expected behavior
[Hooks reference page](https://reactjs.org/docs/hooks-reference.html#timing-of-effects) says that `the function passed to useEffect fires after layout and paint, during a deferred event`. This makes complete sense, but in my example everything works synchronously and browser updates the page when useEffect function is finished. Page also states that `useEffect will fire synchronously before layout and paint when it’s the result of a discrete user input such as a click`, but if it would fire before layout then shouldn't className of test-btn be not updated at the moment of running this function?

**Comment by eps1lon** [comment]
This is intentional behavior for which the docs haven't been updated yet (work is underway). See https://github.com/reactwg/react-18/discussions/128#discussion-3960324 for more in-depth information 

Related: https://github.com/facebook/react/issues/20863

**Comment by NiedziolkaMichal** [comment]
@eps1lon Linked announcement mentiones case of useEffect being triggered by discrete input, such as mouse click. [Docs already were up updated](https://reactjs.org/docs/hooks-reference.html#:~:text=Additionally%2C%20starting%20in%20React%2018,caller%20of%20flushSync.). In my example I am mentioning case of komponent construction and setTimeout which are not such cases. Could you also please take a look at my question about className change?

**Comment by eps1lon** [comment]
Can you convert this to a [codesandbox](https://react.new/) and explain what the expected behavior is you'd want to see and why it is important?

**Comment by NiedziolkaMichal** [comment]
https://codesandbox.io/s/gifted-wright-3e72hp?file=/src/App.js
While investigating, make sure to use F12 console, not codesandbox.io which will have its logs blocked while doing heavy operation.

My understanding from reading docs is that useEffect function should be executed `after layout and paint`, so changes made to the layout should be displayed at this point. I know that exception is useEffect being triggered by mouse clicks, keyboard button clicks and other user interactions, but provided example shows that it's also not the case for timeout and component creation.

My second struggle is with statement `useEffect will fire synchronously before layout and paint when it’s the result of a discrete user input such as a click`. Provided example shows that when useEffect is fired, changes to DOM were already applied, so reflow was already done.

It's important for me only for the purpose of understanding, not because this behavior is problematic.

**Comment by eps1lon** [comment]
> My understanding from reading docs is that useEffect function should be executed after layout and paint, so changes made to the layout should be displayed at this point. 

Yes, the docs are incomplete. As the linked working group issue says, in certain situations it will fire synchronously (like `useLayoutEffect`) and thus after layout but before paint.
